### PR TITLE
Add ceph storage tests.

### PR DESCRIPTION
**Pull request by rbian**

### CODE REVIEW

File: libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg [hunk 1]
```diff
@@ -0,0 +1,100 @@
+- virtual_disks.ceph:
+    type = virtual_disks_ceph
+    take_regular_screendumps = "no"
+    start_vm = "no"
+    disk_source_protocol = "rbd"
+    disk_source_pool= "autotest"
+    disk_source_host = "10.66.4.147 10.66.5.129 10.66.4.150"
+    disk_source_port = "6789 6789 6789"
```
Comment (kylazhang) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg]: It is better not to give internal ip and port in config file

File: libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    disk_source_name = "autotest/rbd.img"
+    disk_format = "raw"
+    mon_host = "10.66.4.147"
```
Comment (kylazhang) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg]: As above

File: libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            unsupported_msg = "this function is not supported by the connection driver: storage pool does( not)?(n't)? support volume"
+            variants:
+                - create_as:
+                - create_xml:
+                    create_by_xml = "yes" 
+    variants:
+        - with_auth:
+            client_name = "client.admin"
+            client_key = "AQBQBjdV0KeiJhAAgMdVun7jybi5QNWP5R1rFQ=="
+            auth_key = "AQCtgYJV8H0uExAAv+mEPw+T6XQ9KBOqpWcX2Q=="
```
Comment (kylazhang) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg]: Also here should be an example String instead of a real key value.

Comment (rbian) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_ceph.cfg]: ok, I will replace all server info with example string. Thanks.

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 4]
```diff
@@ -0,0 +1,667 @@
...
+        if create_by_xml:
+            volxml = vol_xml.VolXML()
+            v_xml = volxml.new_vol(**vol_params)
+            v_xml.xmltreefile.write()
+            ret = virsh.vol_create(pool_name, v_xml.xml, **virsh_dargs)
+            libvirt.check_exit_status(ret)
+        else:
+            pv = libvirt_storage.PoolVolume(pool_name)
+            if not pv.create_volume(vol_name, '2G'):
+                raise error.TestFail("Failed to create storage volume")
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: What do you think if we move pool/vol create part to ```PoolVolumeTest``` in virttest/utils_test/libvirt.py

**Comment by rbian** [comment]
Test results:

16:45:56 (1/13) virtual_disks.ceph.with_auth.network_type.default  Result: PASS 971.91 s
17:02:09 (2/13) virtual_disks.ceph.with_auth.network_type.conf_file  Result: PASS 130.68 s
17:04:21 (3/13) virtual_disks.ceph.with_auth.network_type.iothread  Result: PASS 22.36 s
17:04:44 (4/13) virtual_disks.ceph.with_auth.domain_operation.save_restore  Result: PASS 138.46 s
17:07:04 (5/13) virtual_disks.ceph.with_auth.domain_operation.snapshot.disk_mem  Result: PASS 130.51 s
17:09:16 (6/13) virtual_disks.ceph.with_auth.domain_operation.snapshot.disk_only  Result: PASS 129.66 s
17:11:27 (7/13) virtual_disks.ceph.with_auth.domain_operation.snapshot.internal  Result: PASS 125.22 s
17:13:33 (8/13) virtual_disks.ceph.with_auth.domain_operation.blockcopy  Result: PASS 143.99 s
17:15:58 (9/13) virtual_disks.ceph.with_auth.disk_attach.attach_device  Result: PASS 61.55 s
17:17:01 (10/13) virtual_disks.ceph.with_auth.disk_attach.disk_shareable  Result: PASS 191.90 s
17:20:15 (11/13) virtual_disks.ceph.with_auth.disk_attach.attach_with_snapshot  Result: PASS 80.84 s
17:21:37 (12/13) virtual_disks.ceph.with_auth.disk_pool.create_as  Result: PASS 61.32 s
17:22:39 (13/13) virtual_disks.ceph.with_auth.disk_pool.create_xml  Result: PASS 41.93 s
*  13 cases PASS
Recording test results
Finished: SUCCESS

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 5]
```diff
@@ -0,0 +1,682 @@
+import os
+import re
+import time
+import logging
+from autotest.client.shared import error
+from autotest.client import utils
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: Now we using avocado.utils.process to run cmd

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+from virttest import virsh
+from virttest import aexpect
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: And aexpect is a independent repo now

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+            hosts.append("%s:%s" % (host, port))
+        with open(disk_src_config, 'w') as f:
+            f.write(conf_str + ','.join(hosts) + '\n')
+
+    def create_pool():
+        """
+        Define and start a pool.
+
+        :param p_name. Pool name.
+        :param p_type. Pool type.
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: Well, seems there's no these two param here.
And, other funs below also have this problem.

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+                    # The command can be successfull, but error
+                    # messges can be found from the command output
+                    return False
+                else:
+                    return True
+
+            # Other errors
+            if s != 0:
+                return False
+            return True
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: For read_only, the comment and the return logic is not match. I think you should clean this return part here

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+
+                logging.debug("Secret xml: %s", sec_xml)
+                ret = virsh.secret_define(sec_xml.xml)
+                libvirt.check_exit_status(ret)
+
+                secret_uuid = re.findall(r".+\S+(\ +\S+)\ +.+\S+",
+                                         ret.stdout)[0].lstrip()
+                logging.debug("Secret uuid %s", secret_uuid)
+                if secret_uuid is None:
+                    raise error.TestNAError("Failed to get secret uuid")
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: We should raise SKIP as early as possible

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+                                             **virsh_dargs)
+                libvirt.check_exit_status(ret)
+
+            # Delete the disk if it exists
+            cmd = ("rbd -m {0} {1} info {2} && rbd -m {0} {1} rm "
+                   "{2}".format(mon_host, key_opt, disk_src_name))
+            #utils.system(cmd, ignore_status=True)
+        else:
+            logging.warning("ceph-common couldn't be installed,"
+                            " you may meet error on this test")
```
Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: If this package not installed, is there any chance that the case can run successfully? if not, we should raise error here.

**Comment by rbian** [comment]
Depends on https://github.com/avocado-framework/avocado-vt/pull/337

**Comment by will-Do** [comment]
What's status of this pr?

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 8 above)
Comment (rbian) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: Fixed the messages.

File: libvirt/tests/src/virtual_disks/virtual_disks_ceph.py (hunk 9 above)
Comment (rbian) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: What's the problem here?

Comment (kylazhang) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: @will-Do I also do not see the problem. What's your consideration?

Comment (will-Do) [inline libvirt/tests/src/virtual_disks/virtual_disks_ceph.py]: @rbian @kylazhang 
Just a tip that fir these 'TestNAError' cases, we should raise them as early as possible to save the case run time

**Comment by will-Do** [comment]
Let's merge this case first, any update or fix should raise a new patch/

**Pull request merged**